File: leetcode/coin_change.cpp
/* Author Ahmed Anwar
 * Problem Link https://leetcode.com/problems/coin-change/
 */
#include <bits/stdc++.h>

using namespace std;
typedef unsigned long long ull;
typedef signed long long sll;

int solve(vector<int> &coins, vector<int> &dp, int left) {
    if (dp[])
}

int coinChange(vector<int> &coins, int amount) {
    sort(coins.rend(), coins.rbegin());
    vector<int> dp(coins.size(), 1);
    return solve(coins, dp, amount);
}


int main()
{
    int t;// = 1;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
    }
    return 0;
}

File: leetcode/coin_change.py
'''
Author Ahmed Anwar
Problem Link https://leetcode.com/problems/coin-change/
'''

class Solution:
    def coinChange(self, coins: list[int], amount: int) -> int:
        dp = [amount+1] * (amount+1)
        dp[0] = 0
        for a in range(amount+1):
            for c in coins:
                if (a-c >= 0):
                    dp[a] = min(dp[a], 1+dp[a-c])

        return dp[amount] if dp[amount] != amount + 1 else -1

m = Solution()
print(m.coinChange([1, 2, 5], 11))
